### DISCUSSION TITLE
Failing to execute successive Dynamic Task Group Mapping

### DISCUSSION THREAD

**Discussion by pmercatoris** [apache/airflow#29618] [General]
### Apache Airflow version

2.5.1

### What happened

I am trying to map a task group within another task group which is already mapped. However, when launching the dag, all tasks of the containing task group successfully finish. 
However, the following occurs:  

- The contained task group is not mapped and has a state of  `upstream failed`. 
![image](https://user-images.githubusercontent.com/8655813/216397932-b07c623b-f3ca-45c5-8a52-6bd7911fe538.png)
![image](https://user-images.githubusercontent.com/8655813/216398586-6f8ebb2b-6b99-4df5-898e-1356c66f73be.png)

- The graph UI doesn't show after launching the dag as in https://github.com/apache/airflow/issues/29287 
![image](https://user-images.githubusercontent.com/8655813/216398113-767abd6a-07e6-499b-8f59-4316f9edd695.png)

### What you think should happen instead

I would expect the print_group and print task to start as soon as 1 of the load task finishes

### How to reproduce

I am currently using the docker-compose of the version 2.5.1
`FROM apache/airflow:2.5.1-python3.10`

```python
import pendulum
from airflow.decorators import dag, task, task_group

@task
def get_symbols():
    res = [('A', 1, 111), ('B', 2, 222)]
    return res


@task
def print(symbol_info, data_interval_end=None):
    # Do some work...
    print(symbol_info)
    return symbol_info

@task_group()
def print_group(symbol):
    return print(symbol_info=symbol)


@task
def extract(symbol_info, data_interval_end=None):
    # Do some work...
    return symbol_info
@task
def transform(symbol_info, data_interval_end=None):
    # Do some work...
    return symbol_info

@task
def load(symbol_info, data_interval_end=None):
    # Do some work...
    return 2*[symbol_info]

@task_group
def etl_tg(symbol):
    raw_symbols_data = extract(symbol_info=symbol)
    clean_symbols_data = transform(symbol_info=raw_symbols_data)
    loaded_symbols = load(symbol_info=clean_symbols_data)
    return print_group.expand(symbol=loaded_symbols)

@dag(
    dag_id=f"task_group_mapping_2",
    tags=["sandbox"],
    schedule=None,
    start_date=pendulum.datetime(2023, 1, 1, tz="UTC"),
    catchup=False,
    max_active_runs=1,
)
def etl_dag():

    # DAG
    symbols = get_symbols()
    etl_tg.expand(symbol=symbols)

etl_dag() 
```

### Operating System

Ubuntu 20.04.5 LTS

### Deployment

Docker-Compose

### Are you willing to submit PR?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)

**Comment by hussein-awala** [comment]
Operator and task group expansion in an expanded task group is not supported yet. We raise [NotImplementedError](https://github.com/apache/airflow/blob/main/airflow/models/mappedoperator.py#L309-L309) exception for the operators but I am not sure if it is raised in your case.
 
To implement the mapped tasks, we have added a new column `map_index` which represent the id of each mapped task, and we set it to `-1` for the normal tasks. Supporting the nested expansion is not possible with the same logic, because we need to add a new column for each level, and the number of possible levels is infinite in this case (nested expansion in nested expansion ...), so I don't think we will support it soon.

**Reply by badrisrinivas9** [reply]
Can we have map_index in the task id. Something like group1[0].group2.task3 and then we can reuse the map_index column to specify if its normal task under last group it has expanded and go on for the nested expansion. 
Task_id				map_index
task1			               -1
group1[0].task2			0
group1[1].task2			1
group1[0].group2.task3		-1
group1[0].group2.task4		0
group1[0].group2.task4		1
group1[1].group2.task3		-1
group1[1].group2.task4		0
group1[1].group2.task4		1

in group1 is expanded then in group2 task 4 is expanded. Just checking if it's feasible?

**Comment by potiuk** [comment]
as @hussein-awala explained, this is not supported yet.

**Reply by mis98zb** [reply]
May I know why this is not supported?
I met the same issue and want to fix it.

Is it because the map_index in task_instance cannot support nested mapping?
How about we change it from int to varchar, so as we can support the nest mapping index like '3,3,2'?

And, to trace the nest stack, the `prefix_group_id` option may be forced in task_group.

**Reply by potiuk** [reply]
Because nobody implemented it. If you feel like it - feel free.

**Reply by mis98zb** [reply]
I have a question on the task_concurrencty_dep.
Is it actually used?
I searched the whole source code folder, only when ScheduleJobRunner schedules dag run, the xxxx dep will be called. However, the task_concurrencty_dep is not included in the dep list.
And, there are several DEPS list defined in ti_deps/dependencies_deps.py, but I don't find any code uses the lists.
Has this been deprecated?

Thanks & Best Regards.

**Reply by potiuk** [reply]
Looking at the code it's an automatically added dependency (not ment to be used by users) and it is added automatically when some of the `max_active` kwargs are used. Here is copy of the test that tests it: 

```
class TestTaskConcurrencyDep:
    def _get_task(self, **kwargs):
        return BaseOperator(task_id="test_task", dag=DAG("test_dag"), **kwargs)

    @pytest.mark.parametrize(
        "kwargs, num_running_tis, is_task_concurrency_dep_met",
        [
            ({}, None, True),
            ({"max_active_tis_per_dag": 1}, 0, True),
            ({"max_active_tis_per_dag": 2}, 1, True),
            ({"max_active_tis_per_dag": 2}, 2, False),
            ({"max_active_tis_per_dagrun": 2}, 1, True),
            ({"max_active_tis_per_dagrun": 2}, 2, False),
            ({"max_active_tis_per_dag": 2, "max_active_tis_per_dagrun": 2}, 1, True),
            ({"max_active_tis_per_dag": 1, "max_active_tis_per_dagrun": 2}, 1, False),
            ({"max_active_tis_per_dag": 2, "max_active_tis_per_dagrun": 1}, 1, False),
            ({"max_active_tis_per_dag": 1, "max_active_tis_per_dagrun": 1}, 1, False),
        ],
    )
    def test_concurrency(self, kwargs, num_running_tis, is_task_concurrency_dep_met):
        task = self._get_task(start_date=datetime(2016, 1, 1), **kwargs)
        dep_context = DepContext()
        ti = Mock(task=task, execution_date=datetime(2016, 1, 1))
        if num_running_tis is not None:
            ti.get_num_running_task_instances.return_value = num_running_tis
        assert TaskConcurrencyDep().is_met(ti=ti, dep_context=dep_context) == is_task_concurrency_dep_met
```

**Reply by mis98zb** [reply]
Thank you for the help!

I still have question: 
Since it is checked during scheduling the TIs (`ScheduleJobRunner._schedule_dag_run()`), why it is checked again when queue these selected TIs (`ScheduleJobRunner._executable_task_instances_to_queued()`)?
Is this duplicated?

```mermaid
sequenceDiagram

ScheduleJobRunner ->>+ ScheduleJobRunner: _execute()/_run_scheduler_loop()/_do_scheduling()
ScheduleJobRunner ->>+ ScheduleJobRunner: _start_queued_dagruns()
ScheduleJobRunner ->> ScheduleJobRunner: _get_next_dagruns_to_examine(QUEUED)
ScheduleJobRunner ->>- DagRun: notify_dagrun_state_changed()
ScheduleJobRunner ->> ScheduleJobRunner: _get_next_dagruns_to_examine(RUNNING)
ScheduleJobRunner ->>+ ScheduleJobRunner: _schedule_all_dag_runs()
ScheduleJobRunner ->>+ ScheduleJobRunner: _schedule_dag_run()
ScheduleJobRunner ->>+ DagRun: update_state() //schedule_tasks()
DagRun ->>+ DagRun: task_instance_scheduling_decisions()
DagRun ->>+ DagRun: _get_ready_tis()
DagRun ->>+ TaskInstance: are_dependencies_met()
TaskInstance ->>+ TaskInstance: get_failed_dep_statuses()
TaskInstance ->>+ xxxxxDep: BaseTIDep::get_dep_statuses()
xxxxxDep ->>- xxxxxDep: _get_dep_statuses()
deactivate TaskInstance
deactivate TaskInstance
DagRun ->>+ DagRun: _expand_mapped_task_if_needed()
DagRun ->>- xxxxOperator: expand_mapped_task()
deactivate DagRun
deactivate DagRun
deactivate DagRun
ScheduleJobRunner ->>- DagRun: schedule_tis()
deactivate ScheduleJobRunner
ScheduleJobRunner ->> ScheduleJobRunner: _send_dag_callbacks_to_processor() 
ScheduleJobRunner ->>+ ScheduleJobRunner: _critical_section_enqueue_task_instances()
ScheduleJobRunner ->>+ ScheduleJobRunner: _executable_task_instances_to_queued()
ScheduleJobRunner ->> ScheduleJobRunner: query scheduled task instances to examine
ScheduleJobRunner ->>+ ScheduleJobRunner: __get_concurrency_maps()
ScheduleJobRunner ->>- ConcurrencyMap: from_concurrency_map()
deactivate ScheduleJobRunner
ScheduleJobRunner ->>- ScheduleJobRunner: _enqueue_task_instances_with_queued_state()
deactivate ScheduleJobRunner
```

**Reply by potiuk** [reply]
Scheduling happens first (creating dag run entries in the table) and then scheduler schedules tasks. You can find a talk https://www.youtube.com/watch?v=DYC4-xElccE  from Airflow Summit if you want to get a deep dive how scheduler works

**Reply by mis98zb** [reply]
Thank you for the information.
But with the link https://www.youtube.com/watch?v=DYC4-xElccE I only see a 30s advertising video....
Is this link correct?

**Reply by potiuk** [reply]
Not sure. Try to search "deep dive airflow scheduler" - maybe it was a mobile copy&paste, but you will fnd it 

https://www.youtube.com/watch?v=DYC4-xElccE

I try to copy & paste it again, maybe it will work - if not, just search for it.
